packages/grid/x-data-grid/src/hooks/utils/useGridSelector.ts

```diff
@@ -2,31 +2,65 @@ import * as React from 'react';
 import { GridApiCommon } from '../../models/api/gridApiCommon';
 import { OutputSelector } from '../../utils/createSelector';
 import { buildWarning } from '../../utils/warning';
+import { fastShallowCompare } from '../../utils/fastShallowCompare';
+
+const stateNotInitializedWarning = buildWarning([
+  'MUI: `useGridSelector` has been called before the initialization of the state.',
+  'This hook can only be used inside the context of the grid.',
+]);
+
+const EMPTY = [] as unknown[];
 
 function isOutputSelector<Api extends GridApiCommon, T>(
   selector: any,
 ): selector is OutputSelector<Api['state'], T> {
   return selector.acceptsApiRef;
 }
 
-const stateNotInitializedWarning = buildWarning([
-  'MUI: `useGridSelector` has been called before the initialization of the state.',
-  'This hook can only be used inside the context of the grid.',
-]);
+export function applySelector<Api extends GridApiCommon, T>(
+  apiRef: React.MutableRefObject<Api>,
+  selector: ((state: Api['state']) => T) | OutputSelector<Api['state'], T>,
+) {
+  if (isOutputSelector<Api, T>(selector)) {
+    return selector(apiRef);
+  }
+  return selector(apiRef.current.state);
+}
+
+export const defaultCompare = Object.is; // XXX: Do we need to polyfill?
+export const shallowCompare = fastShallowCompare;
 
 export const useGridSelector = <Api extends GridApiCommon, T>(
   apiRef: React.MutableRefObject<Api>,
   selector: ((state: Api['state']) => T) | OutputSelector<Api['state'], T>,
+  equals: (a: T, b: T) => boolean = defaultCompare,
 ) => {
   if (process.env.NODE_ENV !== 'production') {
     if (!apiRef.current.state) {
       stateNotInitializedWarning();
     }
   }
 
-  if (isOutputSelector<Api, T>(selector)) {
-    return selector(apiRef);
-  }
+  const selectorRef = React.useRef<typeof selector>();
+  const [state, setState] = React.useState<T>(
+    (selectorRef.current ? null : applySelector(apiRef, selector)) as T,
+  );
```

Comment:
The current implementation doesn't re-allocate the `() => applySelector(...)` initialization closure on every render.

For both this comment and the one before, I'm aware those are micro-optimizations. I generally prefer to focus on readability first, but sometimes when a function is called a lot I trade some readability for some performance. I've run a few examples, on a small datagrid with compact density there are around 300 `useGridSelector` hooks mounted at all times. On a fullscreen (laptop screen) datagrid, there are around 1400 of them. Those are numbers where I kinda wanna start trading off readability. Lmk what you think.

Reply:
> Yes. The current implementation doesn't re-allocate the () => applySelector(...) initialization closure on every render.

I didn't even think about it as optimization 🤯

> I've run a few examples, on a small datagrid with compact density there are around 300 useGridSelector hooks mounted at all times. On a fullscreen (laptop screen) datagrid, there are around 1400 of them. Those are numbers where I kinda wanna start trading off readability.

Do you have any perf measurements for that? I'm curious to learn how impactful this is 🙂

Reply:
> I didn't even think about it as optimization

Lol yeah I spent my last year doing low-level programming, I keep seeing allocations everywhere in javascript and it drives me crazy.

> Do you have any perf measurements for that?

No measurements :( Memory allocations are fast enough for closures that this won't appear in the profiles, what's slow is the GC pauses. And we can't measure those, so we can't really benchmark this properly.

In other cases, like filtering performance, it's easier to see the impact of allocations because there is much more upfront memory IO (i.e. reading all the values for the `cellParams` field, and writing those to the new object). Also the fact that filtering runs all its computations in a tight loop allows GC pauses to happen during those computation, unlike in event handlers where the engine can run the GC later.

Reply:
Could you add a comment to clarify why the initializer function is not used here?

Reply:
Done